// oauth2-api-azure/lib/middleware/server-config/HttpErrorHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class HttpErrorHandler {
    static create(status, message = "") {
        return {
            status,
            message,
        };
    }
    // 400 Bad Request
    static BadRequest(message) {
        return HttpErrorHandler.create(400, message);
    }
    // 401 Unauthorized Request
    static Unauthorized(message) {
        return HttpErrorHandler.create(401, message || "Access denied");
    }
    // 403 Forbidden Request
    static Forbidden(message) {
        return HttpErrorHandler.create(403, message);
    }
    // 404 Not Found
    static NotFound(message) {
        return HttpErrorHandler.create(404, message);
    }
    // 405 Method not allowed
    static MethodNotAllowed(message) {
        return HttpErrorHandler.create(405, message);
    }
    // 406 Not Acceptable
    static NotAcceptable(message) {
        return HttpErrorHandler.create(406, message);
    }
    // 409 Conflict
    static Conflict(message) {
        return HttpErrorHandler.create(409, message);
    }
    // 423 Locked
    static Locked(message) {
        return HttpErrorHandler.create(423, message);
    }
    // 424 Failed Dependency
    static FailedDependency(message) {
        return HttpErrorHandler.create(424, message);
    }
    // 500 Internal Server Error
    static InternalServerError(message) {
        return HttpErrorHandler.create(500, message);
    }
    // 503 Service unavailable
    static ServiceUnavailable(message) {
        return HttpErrorHandler.create(503, message);
    }
}
exports.HttpErrorHandler = HttpErrorHandler;
